ipcc-key-value/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+//! Utilities for key/value pairs passed from the control plane to the SP
+//! (through MGS) to the host (through the host/SP uart) via IPCC.
+
+use serde::Deserialize;
+use serde::Serialize;
+
+/// Supported keys.
+///
+/// The numeric values of these keys should match the definitions in the SP
+/// `host-sp-comms` task.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(u8)]
+pub enum Key {
+    /// Always responds `"pong"`.
+    Ping = 0,
+    /// The value should be an encoded [`InstallinatorImageId`].
+    InstallinatorImageId = 1,
+}
+
+/// Description of the images `installinator` needs to fetch from a peer on the
+/// bootstrap network during sled recovery.
+// TODO this may need more or different fields
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
+)]
+pub struct InstallinatorImageId {
+    /// SHA-256 hash of the host OS image to fetch.
+    pub host_os: [u8; 32],
+    /// SHA-256 hash of the control plane image to fetch.
+    pub control_plane: [u8; 32],
+}
+
+// TODO the existence of these methods is a _little_ dubious:
+// `InstallinatorImageId` is `pub` and implements `Serialize + Deserialize`, so
+// any caller could choose to serialize/deserialize an instance with any
+// serde serializer/deserializer already. However, in our case MGS will be
+// serializing and installinator will be deserializing, so I think it makes
+// sense that both of those can call these functions to keep the choice of
+// encoding in one place.
+impl InstallinatorImageId {
+    pub fn serialize(&self) -> Vec<u8> {
```

Comment:
Should we ensure that serialization always results in a value smaller than 512 bytes?

Reply:
Hmm good question! My gut feeling is no, because that limit is imposed by the SP, and we could decide to raise it - duplicating the constant here seems overly restrictive. But this isn't a super strong feeling - could easily be talked out of it.